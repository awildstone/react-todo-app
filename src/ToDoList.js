import React, { useState } from 'react';
import ToDo from './ToDo';
import ToDoListForm from './ToDoListForm';
import './ToDoList.css';

const ToDoList = () => {

    const [ toDos, setToDos ] = useState([]);

    const addToDo = (newToDo) => {
        setToDos(toDos => [...toDos, newToDo]);
    }

    const removeToDo = (toDoId) => {
        setToDos(toDos.filter(toDo => toDo.id !== toDoId));
    }

    const editToDo = (toDoId, data) => {
        setToDos(toDos.map(toDo => {
            if (toDo.id === toDoId) {
                toDo.toDo = data;
            }
            return toDo;
        }))
    }

    return (
        <div className="list">
            <h1>ToDo List</h1>
            <ToDoListForm addToDo={ addToDo } />
            <div>
                {toDos.map(({ id, task }) => 
                    <ToDo 
                        key={id} 
                        id={id} 
                        task={task}
                        removeToDo={removeToDo}
                        editToDo={editToDo}
                    />
                )}
            </div>
        </div>
    )
}

export default ToDoList;