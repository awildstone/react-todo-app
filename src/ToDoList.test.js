import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import ToDoList from './ToDoList';

function addTodo(toDoList, task='my test task') {
    const toDoInput = toDoList.getByLabelText('Task:');
    fireEvent.change(toDoInput, { target: { value: task }});
    const addBtn = toDoList.getByText('Add');
    fireEvent.click(addBtn);
}

it('renders a ToDoList without crashing', () => {
    render(<ToDoList />);
});

it('matches a ToDoList snapshot', () => {
    const { asFragment } = render(<ToDoList />);
    expect(asFragment()).toMatchSnapshot();
});

it('can add a new todo', () => {
    const toDoList = render(<ToDoList />);
    addTodo(toDoList);

    //expect new todo form is reset
    expect(toDoList.getByLabelText('Task:')).toHaveValue('');
    
    //expect new todo is in the DOM
    expect(toDoList.getByText('my test task')).toBeInTheDocument();
    expect(toDoList.getByText('Mark as completed')).toBeInTheDocument();
    expect(toDoList.getByText('Edit')).toBeInTheDocument();
    expect(toDoList.getByText('X')).toBeInTheDocument();
});

it('can edit a todo', () => {
    const toDoList = render(<ToDoList />);
    addTodo(toDoList);

    //confirm the new todo is in the document
    expect(toDoList.queryByText('my test task')).toBeInTheDocument();

    //click the edit button
    fireEvent.click(toDoList.getByText('Edit'));

    //modify the input
    const toDoInput = toDoList.getByDisplayValue('my test task');
    fireEvent.change(toDoInput, { target: { value: 'new test value' }});
    fireEvent.click(toDoList.getByText('Update'));

    //expect only the new todo text
    expect(toDoList.queryByText('my test task')).not.toBeInTheDocument();
    expect(toDoList.getByText('new test value')).toBeInTheDocument();
});

it('can delete a todo', () => {
    const toDoList = render(<ToDoList />);
    addTodo(toDoList);

    //confirm the new todo is in the document
    expect(toDoList.queryByText('my test task')).toBeInTheDocument();

    //click the delete button
    fireEvent.click(toDoList.getByText('X'));

    //expect the todo is now deleted
    expect(toDoList.queryByText('my test task')).not.toBeInTheDocument();
});

it('can toggle a todo complete', () => {
    const toDoList = render(<ToDoList />);
    addTodo(toDoList);

    //confirm the new todo is in the document
    expect(toDoList.queryByText('my test task')).toBeInTheDocument();

    //click the toggle mark complete button
    fireEvent.click(toDoList.getByText('Mark as completed'));

    //confirm todo is crossed out
    expect(toDoList.queryByText('my test task')).toHaveStyle({'text-decoration': 'line-through'});

});

it('can untoggle a completed todo', () => {
    const toDoList = render(<ToDoList />);
    addTodo(toDoList);

    //confirm the new todo is in the document
    expect(toDoList.queryByText('my test task')).toBeInTheDocument();

    //toggle the todo completed and confirm style is applied
    fireEvent.click(toDoList.getByText('Mark as completed'));
    expect(toDoList.queryByText('my test task')).toHaveStyle({'text-decoration': 'line-through'});

    //click the todo completed again and confirm the style is removed
    fireEvent.click(toDoList.getByText('Mark as completed'));
    expect(toDoList.queryByText('my test task')).not.toHaveStyle({'text-decoration': 'line-through'});
});