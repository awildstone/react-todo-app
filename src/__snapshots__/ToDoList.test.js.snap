// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches a ToDoList snapshot 1`] = `
<DocumentFragment>
  <div
    class="list"
  >
    <h1>
      ToDo List
    </h1>
    <div
      class="new-todo"
    >
      <form>
        <label
          for="task"
        >
          Task: 
        </label>
        <input
          id="task"
          name="task"
          type="text"
        />
        <button
          class="save"
        >
          Add
        </button>
      </form>
    </div>
    <div />
  </div>
</DocumentFragment>
`;
